perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -1207,76 +1197,93 @@ public void synchronizeGroups(PerunSession sess) throws InternalErrorException {
 		private Group group;
 		private long startTime;
 
-		public GroupSynchronizerThread(PerunSession sess, Group group) {
+		public GroupSynchronizerThread(PerunSession sess) {
 			// take only reference to perun
 			this.perunBl = (PerunBl) sess.getPerun();
-			this.group = group;
 			try {
 				// create own session
 				this.sess = perunBl.getPerunSession(pp, new PerunClient());
 			} catch (InternalErrorException ex) {
 				log.error("Unable to create internal session for Synchronizer with credentials {} because of exception {}", pp, ex);
 			}
+			//Default settings of not running thread (waiting for another group)
+			this.group = null;
+			this.startTime = 0;
 		}
 
 		public void run() {
-			//text of exception if was thrown, null in exceptionMessage means "no exception, it's ok"
-			String exceptionMessage = null;
-			//text with all skipped members and reasons of this skipping
-			String skippedMembersMessage = null;
-			//if exception which produce fail of whole synchronization was thrown
-			boolean failedDueToException = false;
+			while (true) {
+				//Set thread to default state (waiting for another group to synchronize)
+				this.setThreadToDefaultState();
+				//text of exception if was thrown, null in exceptionMessage means "no exception, it's ok"
+				String exceptionMessage = null;
+				//text with all skipped members and reasons of this skipping
+				String skippedMembersMessage = null;
+				//if exception which produce fail of whole synchronization was thrown
+				boolean failedDueToException = false;
 
-			try {
-				log.debug("Synchronization thread for group {} has started.", group);
-				// Set the start time, so we can check the timeout of the thread
-				startTime = System.currentTimeMillis();
+				try {
+					//Take anouther group from the queue to synchronize it
+					this.group = GroupsManagerBlImpl.this.takeAnotherGroupFromQueueToBeSynchronized();
```

Comment:
Can return null, which causes NullPointerException on following statement.

Reply:
I suggest replacing the Deque implementation with LinkedBlockingDeque which provides takeFirst() method that blocks instead of returning null if deque is empty